### DISCUSSION TITLE
How to get query params?

### DISCUSSION THREAD

**Discussion by chiralium** [vercel/next.js#44005] [Help]
### Summary

How to get query parameters from new router? (`next/navigation`). New router has been dividev by few new hooks:
- `usePathname` -- to get pathname;
- `useSearchParams` -- to get `get`-search params
- `useRouter` -- that return function for handling routing

I'm currently try to migrate my Next 12 project to Next 13, and having problem:

```typescript
export const useDynamicPathName = (paramKey: string): string => {
    const {query} = useRouter();

    const queryData = useMemo<string[]>(() => {
        return query[paramKey] as string[];
    }, [paramKey, query]);

    return useMemo<string>(() => {
        if (!queryData) {
            return '';
        }a

        return queryData.join('/');
    }, [queryData]);
};
```

my custom hook now is not working, cuz new `useRouter`-hook do not contain query-object. Single place where I having access to page slug for example (`[...slug]`) this is `page.tsx` in component props.

**Comment by hubertlepicki** [comment]
I am just looking at this myself, and it appears that for the server-side pages/components in the Next.js 13 using "app" architecture you are getting `searchParams` prop automatially passed to each page. So, in my case in `app/page.js` I have:

```
export default function Home({searchParams}) {    
  return (    
    <div>    
      <h1>Hello {searchParams.who}</h1>    
    </div>         
  )                                                                                   
};
```

And if I visit http://localhost:3000/?who=Hubert I do see "Hello Hubert". 

Docs: https://beta.nextjs.org/docs/api-reference/file-conventions/page

And for the client components indeed `useSearchParams` is the way to go.
https://beta.nextjs.org/docs/api-reference/use-search-params#searchparams-in-server-components

But then, this fallls apart once I make static build with `npx next build`. It builds the page as staticly rendered page, meaning it no longer works. I was able to force it to be dynamically rendered by exporting `dynamic='force-dynamic'` this way:

```
export const dynamic='force-dynamic';

export default function Home({searchParams}) {
  return (
    <div>
      <h1>Hello {searchParams.who}</h1>
    </div>
  )
};

```

This seems to work in dev and production alike.

**Reply by GabenGar** [reply]
Yeah there are no search params during static build, the server can only access them within the dynamic context.

**Reply by hubertlepicki** [reply]
@GabenGar indeed, that's why forcing dynamic rendering comes to play:

```
export const dynamic='force-dynamic';
```

When the build is being made it appears the little "lambda" or "epsilon" symbols are being printed next to the pages that are dynamic, and that's what you want to see if you want to use the search params on such pages.

**Reply by OkashaTanoli** [reply]
is there any way to generate this type of static page in next js 13 using generateStaticParams the page URL should be like that 
http://localhost:3000/tracks/wmd?quarter=4

**Reply by cachho** [reply]
> is there any way to generate this type of static page in next js 13 using generateStaticParams the page URL should be like that http://localhost:3000/tracks/wmd?quarter=4

interested in this too, how else are you handling pre-generated pagination?

**Reply by GabenGar** [reply]
By making pagination the part of the path instead of a search query.

**Reply by cachho** [reply]
> By making pagination the part of the path instead of a search query.

yeah but no one does that.

**Reply by OkashaTanoli** [reply]
> By making pagination the part of the path instead of a search query.

can you pleases explain this with example

**Reply by cachho** [reply]
> > By making pagination the part of the path instead of a search query.
> 
> can you pleases explain this with example

He means using `shop/page/2` instead of `/shop?page=2` (which is what I want). So this is using dynamic routes, which is well documented with and without AppDir. Then you can use generateStaticParams to generate static pages at build time.

**Reply by OkashaTanoli** [reply]
> > > By making pagination the part of the path instead of a search query.
> > 
> > 
> > can you pleases explain this with example
> 
> He means using `shop/page/2` instead of `/shop?page=2` (which is what I want). So this is using dynamic routes, which is well documented with and without AppDir. Then you can use generateStaticParams to generate static pages at build time.

ok thank u soo much

**Reply by GabenGar** [reply]
Ofc when directly "ported" like this it looks stupid. But to have a pagination as a part of pathname you have to think about your pages. Basically you'll have to have separate pages for a list of items and item's details.
So for example you'd have `/shop/products/[page]` for the list and `/shop/product/[product_id]` for the details on an item in the list.

**Comment by nos-nart** [comment]
@chiralium Have you found the way to get around this? Am using next^13.1.6 still cannot obtain the `query` from **useRouter**

**Reply by smartercow** [reply]
You have to get it from server component like this:

The code below is in /app/products/[slug]/page.tsx
🛑 No 'use client'
```
export default function ProductPage({params}: {params: { slug: string }}) {
  return (
    <>{params.slug}</>
  );
}
```
The code ☝️ is equivalence to Next.js 12:
/pages/products/[slug]/index.tsx
```
import { useRouter } from 'next/router';

const router = useRouter();

const slug = router.query.slug

  return (
    <>{slug}</>
  );
```

**Reply by chiralium** [reply]
> @chiralium Have you found the way to get around this? Am using next^13.1.6 still cannot obtain the `query` from **useRouter**

No :( We stay on Next 12

**Reply by hifilorau** [reply]
I'm accesing params from server side pages.js component with just params in next13 and it's working fine. 

```
export default function({params}){
 
  console.log('SEARCH PARAMS',params)
  return ( 
    <BotContainer />
  )
}
```

**Comment by uragecz** [comment]
What about if i wanna get params in child component that isn't client component? I don't wanna keep passing down this prop from very top(page) to this child because it's deep, and i also don't wanna transfer this component into client just to get param from url. Any other ways how to get it ?

**Reply by GabenGar** [reply]
It's very unlikely you are going to pass params to nested level components without parsing/validating/error handling first. So pass the values to the arguments just like in any other part of context-dependant procedural code.

**Comment by viv-kumar** [comment]
![Screenshot (180)](https://github.com/vercel/next.js/assets/108941852/76b80159-76d1-47d9-a1bf-f5a8e41a65a8)
please tell how to get query in the GET function here

**Reply by nykoos** [reply]
Did you find out how to do it?

**Reply by OkashaTanoli** [reply]
request: Request
const url = new URL(request.url);
let searchParams = new URLSearchParams(url.search);
let limit= searchParams.get('limit')

// // // 

request: NextRequest
const limit = req.nextUrl.searchParams.get("limit")

**Reply by devjiwonchoi** [reply]
@viv-kumar @Nykoos 
```ts
req: NextRequest

req.nextUrl.searchParams

```
